gdeep/trainer/regularizer.py

```diff
@@ -0,0 +1,434 @@
+            0: positive filtration, 1: negative filtration
+    Returns:
+        coordinates:
+            The coordinates of the critical points that
+            contribute to the complexity of the decision boundary.
+    """
+    res = _evaluate_model_on_grid(dummy_loader, net)
+    if ind == 1:  # negative filtration 'class 0 probability'
+        res = 1 - res
+    # Get the persistent pairs of the linear surrogate function
+    critical_values, indices_of_critical_points = _get_persistent_pairs(
+        res, connections, expansion
+    )
+
+    """ Next we filter out the persistent pairs that contribute to
+    the complexity of the decision boundary:
+    These are the ones that are active at the db: so born before it and die after it.
+    Recall f: X -> [0,1], maps data to class 1 probability (class 0 probability)
+    
+    In that case the decision boundary: f^{-1}(1/2)
+    
+    We can simplify the boundary either by pushing the homology generator above it (delay its birth)
+    or pushing the annihilator below it (expedite its death). We should do whichever is more economical,
+    i.e., closer to the cutoff (1/2). Next we do just that.
+    
+    In the following code variable name containing 'rel_inds' is used to refer to indices with relative to
+    the critical points. (i.e. rel_inds = 0 is the first critical pair)
+    'inds' without this qualifier are indices with respect to the coordinates (inds = 0: first vertex of coords)
+    """
+    # signed distance of the critical values (of the critical pairs) from the db:
+    signed_distance_from_db = critical_values - 0.5
+    # pairs of critical points on different sides of the db: i.e. signed distance XOR negative (or equiv. product<0)
+    db_pairs_rel_inds = (
+        signed_distance_from_db[:, 0] * signed_distance_from_db[:, 1]
+    ) < 0  # column XOR Negative
+    # for each pair of critical points, decide if birth or death is closer to the db:
+    columns = np.argmin(
+        abs(signed_distance_from_db[db_pairs_rel_inds, :]), 1
+    )  # argmin of columns
+    # the indices of the vertices that are on either side of db, expressed in original coords indices:
+    db_pair_inds = indices_of_critical_points[db_pairs_rel_inds, :]
+    # Next we loop through the selected pairs to see which one we push: birth or death, and record the index of that vertex:
+    indices = np.zeros(db_pair_inds.shape[0])
+    for i in range(db_pair_inds.shape[0]):
+        column = columns[i]  #:0 birth, 1: death
+        indices[i] = db_pair_inds[
+            i, column
+        ]  # pick the index of the coordinate accordingly
+    # indices of the selected critical points, in reference to the original coords:
+    indices = indices.astype(int)
+    # return the coordinates, directly digestible by the model
+    coordinates = coords[indices, :].clone().detach().requires_grad_(True)
+    return coordinates
+
+
+class Regularizer:
```

Comment:
In my opinion, the usage of the abstract class `Regularizer` is complicating things more than it simplifies. Both `TopologicalRegularizer` and `TihonovRegularizer` have a very different interface. That's also the reason why the `__init__` method has as argument `**kwargs` and not some properly defined parameters.

The only thing that `TopologicalRegularizer` and `TihonovRegularizer` have in common is that they have a `regularization_penalty` method. Hence, I would build an interface for that. One option would be to define an abstract class only having an *abstract* `regularization_penalty` method:
```
from abc import ABC, abstract_method
class Regularizer(ABC):
    @abstract_method
    def regularization_penalty(self, model: nn.Module) -> Tenor:
        pass
class TopologicalRegularizerData(Regularizer):
    ...

class TihonovRegularizer(Regularizer):
    ...
```
 What I would prefer is to use a protocol instead, then you don't have the inherit from `Regularizer`:
```
from typing import Protocol
class Regularizer:
    def regularization_penalty(self, model: nn.Module) -> Tenor:
        pass

class TopologicalRegularizer:
    ...

class TihonovRegularizer:
    ...
```

Reply:
done (the 2nd approach)